File: StockPilotAI/frontend/src/api/stock.ts
export interface StockBasicInfo {
  current_price: number | null
  market_cap: number | null
  trailing_pe: number | null
  fifty_two_week_high: number | null
  fifty_two_week_low: number | null
}

export interface StockHistoryItem {
  date: string
  open: number
  high: number
  low: number
  close: number
  volume: number
}

export interface StockResponse {
  symbol: string
  basic_info: StockBasicInfo
  history: StockHistoryItem[]
}

interface ApiErrorBody {
  detail?: string | { message?: string }
}

const apiBaseUrl =
  import.meta.env.VITE_API_BASE_URL ?? 'http://localhost:8000/api/v1'

export async function getStock(symbol: string): Promise<StockResponse> {
  const response = await fetch(`${apiBaseUrl}/stock/${encodeURIComponent(symbol)}`)

  if (!response.ok) {
    let message = 'Unable to load market data.'
    try {
      const body = (await response.json()) as ApiErrorBody
      if (typeof body.detail === 'string') {
        message = body.detail
      } else if (body.detail?.message) {
        message = body.detail.message
      }
    } catch {
      // Keep the safe fallback when the upstream response is not JSON.
    }
    throw new Error(message)
  }

  return (await response.json()) as StockResponse
}

File: StockPilotAI/backend/app/schemas/stock.py
from datetime import date

from pydantic import BaseModel, ConfigDict


class StockBasicInfo(BaseModel):
    company_name: str | None = None
    sector: str | None = None
    industry: str | None = None
    current_price: float | None
    market_cap: int | None
    trailing_pe: float | None
    fifty_two_week_high: float | None
    fifty_two_week_low: float | None


class StockHistoryItem(BaseModel):
    date: date
    open: float
    high: float
    low: float
    close: float
    volume: int


class StockResponse(BaseModel):
    model_config = ConfigDict(json_schema_extra={"description": "One year of daily stock data"})

    symbol: str
    basic_info: StockBasicInfo
    history: list[StockHistoryItem]
